File: packages/go/currency_transaction_test.go
package constellation

import (
	"testing"
)

func TestUtilityFunctions(t *testing.T) {
	t.Run("TokenToUnits converts correctly", func(t *testing.T) {
		if TokenToUnits(100.5) != 10050000000 {
			t.Errorf("TokenToUnits(100.5) = %d, want 10050000000", TokenToUnits(100.5))
		}
		if TokenToUnits(0.00000001) != 1 {
			t.Errorf("TokenToUnits(0.00000001) = %d, want 1", TokenToUnits(0.00000001))
		}
		if TokenToUnits(1) != 100000000 {
			t.Errorf("TokenToUnits(1) = %d, want 100000000", TokenToUnits(1))
		}
	})

	t.Run("UnitsToToken converts correctly", func(t *testing.T) {
		if UnitsToToken(10050000000) != 100.5 {
			t.Errorf("UnitsToToken(10050000000) = %f, want 100.5", UnitsToToken(10050000000))
		}
		if UnitsToToken(1) != 0.00000001 {
			t.Errorf("UnitsToToken(1) = %f, want 0.00000001", UnitsToToken(1))
		}
		if UnitsToToken(100000000) != 1.0 {
			t.Errorf("UnitsToToken(100000000) = %f, want 1.0", UnitsToToken(100000000))
		}
	})

	t.Run("TokenDecimals constant", func(t *testing.T) {
		if TokenDecimals != 1e-8 {
			t.Errorf("TokenDecimals = %f, want 1e-8", TokenDecimals)
		}
	})

	t.Run("IsValidDAGAddress validates addresses", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		if !IsValidDAGAddress(keyPair.Address) {
			t.Errorf("IsValidDAGAddress(%s) = false, want true", keyPair.Address)
		}
		if IsValidDAGAddress("invalid") {
			t.Error("IsValidDAGAddress('invalid') = true, want false")
		}
		if IsValidDAGAddress("") {
			t.Error("IsValidDAGAddress('') = true, want false")
		}
	})
}

func TestTransactionCreation(t *testing.T) {
	t.Run("CreateCurrencyTransaction creates valid transaction", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()

		lastRef := TransactionReference{
			Hash:    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			Ordinal: 0,
		}

		tx, err := CreateCurrencyTransaction(
			TransferParams{
				Destination: keyPair2.Address,
				Amount:      100.5,
				Fee:         0,
			},
			keyPair.PrivateKey,
			lastRef,
		)

		if err != nil {
			t.Fatalf("CreateCurrencyTransaction failed: %v", err)
		}
		if tx == nil {
			t.Fatal("Transaction is nil")
		}
		if tx.Value.Source != keyPair.Address {
			t.Errorf("Source = %s, want %s", tx.Value.Source, keyPair.Address)
		}
		if tx.Value.Destination != keyPair2.Address {
			t.Errorf("Destination = %s, want %s", tx.Value.Destination, keyPair2.Address)
		}
		if tx.Value.Amount != 10050000000 {
			t.Errorf("Amount = %d, want 10050000000", tx.Value.Amount)
		}
		if tx.Value.Fee != 0 {
			t.Errorf("Fee = %d, want 0", tx.Value.Fee)
		}
		if tx.Value.Parent != lastRef {
			t.Errorf("Parent mismatch")
		}
		if len(tx.Proofs) != 1 {
			t.Errorf("Proofs length = %d, want 1", len(tx.Proofs))
		}
	})

	t.Run("CreateCurrencyTransaction throws on invalid destination", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		lastRef := TransactionReference{Hash: "a" + string(make([]byte, 63)), Ordinal: 0}

		_, err := CreateCurrencyTransaction(
			TransferParams{Destination: "invalid", Amount: 100, Fee: 0},
			keyPair.PrivateKey,
			lastRef,
		)

		if err != ErrInvalidAddress {
			t.Errorf("Expected ErrInvalidAddress, got %v", err)
		}
	})

	t.Run("CreateCurrencyTransaction throws on same address", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		lastRef := TransactionReference{Hash: "a" + string(make([]byte, 63)), Ordinal: 0}

		_, err := CreateCurrencyTransaction(
			TransferParams{Destination: keyPair.Address, Amount: 100, Fee: 0},
			keyPair.PrivateKey,
			lastRef,
		)

		if err != ErrSameAddress {
			t.Errorf("Expected ErrSameAddress, got %v", err)
		}
	})

	t.Run("CreateCurrencyTransaction throws on amount too small", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()
		lastRef := TransactionReference{Hash: "a" + string(make([]byte, 63)), Ordinal: 0}

		_, err := CreateCurrencyTransaction(
			TransferParams{Destination: keyPair2.Address, Amount: 0.000000001, Fee: 0},
			keyPair.PrivateKey,
			lastRef,
		)

		if err != ErrInvalidAmount {
			t.Errorf("Expected ErrInvalidAmount, got %v", err)
		}
	})

	t.Run("CreateCurrencyTransaction throws on negative fee", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()
		lastRef := TransactionReference{Hash: "a" + string(make([]byte, 63)), Ordinal: 0}

		_, err := CreateCurrencyTransaction(
			TransferParams{Destination: keyPair2.Address, Amount: 100, Fee: -1},
			keyPair.PrivateKey,
			lastRef,
		)

		if err != ErrInvalidFee {
			t.Errorf("Expected ErrInvalidFee, got %v", err)
		}
	})
}

func TestBatchTransactions(t *testing.T) {
	t.Run("CreateCurrencyTransactionBatch creates multiple transactions", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		recipient1, _ := GenerateKeyPair()
		recipient2, _ := GenerateKeyPair()
		recipient3, _ := GenerateKeyPair()

		lastRef := TransactionReference{
			Hash:    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			Ordinal: 5,
		}

		transfers := []TransferParams{
			{Destination: recipient1.Address, Amount: 10},
			{Destination: recipient2.Address, Amount: 20},
			{Destination: recipient3.Address, Amount: 30},
		}

		txns, err := CreateCurrencyTransactionBatch(transfers, keyPair.PrivateKey, lastRef)

		if err != nil {
			t.Fatalf("CreateCurrencyTransactionBatch failed: %v", err)
		}
		if len(txns) != 3 {
			t.Errorf("Batch length = %d, want 3", len(txns))
		}
		if txns[0].Value.Amount != 1000000000 {
			t.Errorf("Transaction 0 amount = %d, want 1000000000", txns[0].Value.Amount)
		}
		if txns[1].Value.Amount != 2000000000 {
			t.Errorf("Transaction 1 amount = %d, want 2000000000", txns[1].Value.Amount)
		}
		if txns[2].Value.Amount != 3000000000 {
			t.Errorf("Transaction 2 amount = %d, want 3000000000", txns[2].Value.Amount)
		}

		// Check parent references are chained
		if txns[0].Value.Parent != lastRef {
			t.Error("Transaction 0 parent mismatch")
		}
		if txns[1].Value.Parent.Ordinal != 6 {
			t.Errorf("Transaction 1 ordinal = %d, want 6", txns[1].Value.Parent.Ordinal)
		}
		if txns[2].Value.Parent.Ordinal != 7 {
			t.Errorf("Transaction 2 ordinal = %d, want 7", txns[2].Value.Parent.Ordinal)
		}
	})
}

func TestTransactionVerification(t *testing.T) {
	t.Run("VerifyCurrencyTransaction validates correct signatures", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()
		lastRef := TransactionReference{
			Hash:    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			Ordinal: 0,
		}

		tx, _ := CreateCurrencyTransaction(
			TransferParams{Destination: keyPair2.Address, Amount: 100, Fee: 0},
			keyPair.PrivateKey,
			lastRef,
		)

		result := VerifyCurrencyTransaction(tx)

		if !result.IsValid {
			t.Error("Transaction should be valid")
		}
		if len(result.ValidProofs) != 1 {
			t.Errorf("ValidProofs length = %d, want 1", len(result.ValidProofs))
		}
		if len(result.InvalidProofs) != 0 {
			t.Errorf("InvalidProofs length = %d, want 0", len(result.InvalidProofs))
		}
	})

	t.Run("VerifyCurrencyTransaction detects invalid signatures", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()
		lastRef := TransactionReference{
			Hash:    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			Ordinal: 0,
		}

		tx, _ := CreateCurrencyTransaction(
			TransferParams{Destination: keyPair2.Address, Amount: 100, Fee: 0},
			keyPair.PrivateKey,
			lastRef,
		)

		// Corrupt the signature
		tx.Proofs[0].Signature = "invalid_signature"

		result := VerifyCurrencyTransaction(tx)

		if result.IsValid {
			t.Error("Transaction should be invalid")
		}
		if len(result.ValidProofs) != 0 {
			t.Errorf("ValidProofs length = %d, want 0", len(result.ValidProofs))
		}
		if len(result.InvalidProofs) != 1 {
			t.Errorf("InvalidProofs length = %d, want 1", len(result.InvalidProofs))
		}
	})
}

func TestMultiSignatureSupport(t *testing.T) {
	t.Run("SignCurrencyTransaction adds additional signature", func(t *testing.T) {
		keyPair1, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()
		recipient, _ := GenerateKeyPair()
		lastRef := TransactionReference{
			Hash:    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			Ordinal: 0,
		}

		// Create transaction with first signature
		tx, _ := CreateCurrencyTransaction(
			TransferParams{Destination: recipient.Address, Amount: 100, Fee: 0},
			keyPair1.PrivateKey,
			lastRef,
		)

		if len(tx.Proofs) != 1 {
			t.Errorf("Initial proofs length = %d, want 1", len(tx.Proofs))
		}

		// Add second signature
		tx, err := SignCurrencyTransaction(tx, keyPair2.PrivateKey)
		if err != nil {
			t.Fatalf("SignCurrencyTransaction failed: %v", err)
		}

		if len(tx.Proofs) != 2 {
			t.Errorf("Proofs length after signing = %d, want 2", len(tx.Proofs))
		}

		// Verify both signatures
		result := VerifyCurrencyTransaction(tx)

		if !result.IsValid {
			t.Error("Transaction should be valid")
		}
		if len(result.ValidProofs) != 2 {
			t.Errorf("ValidProofs length = %d, want 2", len(result.ValidProofs))
		}
		if len(result.InvalidProofs) != 0 {
			t.Errorf("InvalidProofs length = %d, want 0", len(result.InvalidProofs))
		}
	})
}

func TestTransactionHashing(t *testing.T) {
	t.Run("HashCurrencyTransaction produces consistent hashes", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()
		lastRef := TransactionReference{
			Hash:    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			Ordinal: 0,
		}

		tx, _ := CreateCurrencyTransaction(
			TransferParams{Destination: keyPair2.Address, Amount: 100, Fee: 0},
			keyPair.PrivateKey,
			lastRef,
		)

		hash1 := HashCurrencyTransaction(tx)
		hash2 := HashCurrencyTransaction(tx)

		if hash1.Value != hash2.Value {
			t.Error("Hash values should be consistent")
		}
		if len(hash1.Value) != 64 {
			t.Errorf("Hash value length = %d, want 64", len(hash1.Value))
		}
		if len(hash1.Bytes) != 32 {
			t.Errorf("Hash bytes length = %d, want 32", len(hash1.Bytes))
		}
	})

	t.Run("GetTransactionReference creates correct reference", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()
		lastRef := TransactionReference{
			Hash:    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			Ordinal: 0,
		}

		tx, _ := CreateCurrencyTransaction(
			TransferParams{Destination: keyPair2.Address, Amount: 100, Fee: 0},
			keyPair.PrivateKey,
			lastRef,
		)

		ref := GetTransactionReference(tx, 1)

		if ref.Ordinal != 1 {
			t.Errorf("Reference ordinal = %d, want 1", ref.Ordinal)
		}
		if len(ref.Hash) != 64 {
			t.Errorf("Reference hash length = %d, want 64", len(ref.Hash))
		}
	})

	t.Run("EncodeCurrencyTransaction returns string", func(t *testing.T) {
		keyPair, _ := GenerateKeyPair()
		keyPair2, _ := GenerateKeyPair()
		lastRef := TransactionReference{
			Hash:    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			Ordinal: 0,
		}

		tx, _ := CreateCurrencyTransaction(
			TransferParams{Destination: keyPair2.Address, Amount: 100, Fee: 0},
			keyPair.PrivateKey,
			lastRef,
		)

		encoded := EncodeCurrencyTransaction(tx)

		if len(encoded) == 0 {
			t.Error("Encoded transaction should not be empty")
		}
	})
}

File: packages/java/src/test/java/io/constellationnetwork/metagraph/sdk/CurrencyTransactionTest.java
package io.constellationnetwork.metagraph.sdk;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class CurrencyTransactionTest {

    @Nested
    @DisplayName("Utility functions")
    class UtilityFunctions {

        @Test
        @DisplayName("tokenToUnits converts correctly")
        void tokenToUnitsConvertsCorrectly() {
            assertEquals(10050000000L, CurrencyTransaction.tokenToUnits(100.5));
            assertEquals(1L, CurrencyTransaction.tokenToUnits(0.00000001));
            assertEquals(100000000L, CurrencyTransaction.tokenToUnits(1.0));
        }

        @Test
        @DisplayName("unitsToToken converts correctly")
        void unitsToTokenConvertsCorrectly() {
            assertEquals(100.5, CurrencyTransaction.unitsToToken(10050000000L));
            assertEquals(0.00000001, CurrencyTransaction.unitsToToken(1L));
            assertEquals(1.0, CurrencyTransaction.unitsToToken(100000000L));
        }

        @Test
        @DisplayName("TOKEN_DECIMALS constant is correct")
        void tokenDecimalsConstant() {
            assertEquals(1e-8, CurrencyTypes.TOKEN_DECIMALS);
        }

        @Test
        @DisplayName("isValidDagAddress validates addresses")
        void isValidDagAddressValidatesAddresses() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            assertTrue(CurrencyTransaction.isValidDagAddress(keyPair.getAddress()));
            assertFalse(CurrencyTransaction.isValidDagAddress("invalid"));
            assertFalse(CurrencyTransaction.isValidDagAddress(""));
            assertFalse(CurrencyTransaction.isValidDagAddress("DAG"));
        }
    }

    @Nested
    @DisplayName("Transaction creation")
    class TransactionCreation {

        @Test
        @DisplayName("creates valid currency transaction")
        void createsValidCurrencyTransaction() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();

            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            CurrencyTypes.CurrencyTransaction tx = CurrencyTransaction.createCurrencyTransaction(
                new CurrencyTypes.TransferParams(keyPair2.getAddress(), 100.5, 0),
                keyPair.getPrivateKey(),
                lastRef
            );

            assertNotNull(tx);
            assertEquals(keyPair.getAddress(), tx.getValue().getSource());
            assertEquals(keyPair2.getAddress(), tx.getValue().getDestination());
            assertEquals(10050000000L, tx.getValue().getAmount()); // 100.5 * 1e8
            assertEquals(0, tx.getValue().getFee());
            assertEquals(lastRef, tx.getValue().getParent());
            assertEquals(1, tx.getProofs().size());
            assertNotNull(tx.getProofs().get(0).getId());
            assertNotNull(tx.getProofs().get(0).getSignature());
        }

        @Test
        @DisplayName("throws on invalid destination")
        void throwsOnInvalidDestination() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            Exception exception = assertThrows(Types.SdkException.class, () -> {
                CurrencyTransaction.createCurrencyTransaction(
                    new CurrencyTypes.TransferParams("invalid", 100, 0),
                    keyPair.getPrivateKey(),
                    lastRef
                );
            });

            assertTrue(exception.getMessage().contains("Invalid destination address"));
        }

        @Test
        @DisplayName("throws on same source and destination")
        void throwsOnSameAddress() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            Exception exception = assertThrows(Types.SdkException.class, () -> {
                CurrencyTransaction.createCurrencyTransaction(
                    new CurrencyTypes.TransferParams(keyPair.getAddress(), 100, 0),
                    keyPair.getPrivateKey(),
                    lastRef
                );
            });

            assertTrue(exception.getMessage().contains("Source and destination addresses cannot be the same"));
        }

        @Test
        @DisplayName("throws on amount too small")
        void throwsOnAmountTooSmall() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            Exception exception = assertThrows(Types.SdkException.class, () -> {
                CurrencyTransaction.createCurrencyTransaction(
                    new CurrencyTypes.TransferParams(keyPair2.getAddress(), 0.000000001, 0),
                    keyPair.getPrivateKey(),
                    lastRef
                );
            });

            assertTrue(exception.getMessage().contains("Transfer amount must be greater than 1e-8"));
        }

        @Test
        @DisplayName("throws on negative fee")
        void throwsOnNegativeFee() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            Exception exception = assertThrows(Types.SdkException.class, () -> {
                CurrencyTransaction.createCurrencyTransaction(
                    new CurrencyTypes.TransferParams(keyPair2.getAddress(), 100, -1),
                    keyPair.getPrivateKey(),
                    lastRef
                );
            });

            assertTrue(exception.getMessage().contains("Fee must be greater than or equal to zero"));
        }
    }

    @Nested
    @DisplayName("Batch transactions")
    class BatchTransactions {

        @Test
        @DisplayName("creates multiple transactions")
        void createsMultipleTransactions() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair recipient1 = Wallet.generateKeyPair();
            Types.KeyPair recipient2 = Wallet.generateKeyPair();
            Types.KeyPair recipient3 = Wallet.generateKeyPair();

            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 5
            );

            List<CurrencyTypes.TransferParams> transfers = Arrays.asList(
                new CurrencyTypes.TransferParams(recipient1.getAddress(), 10, 0),
                new CurrencyTypes.TransferParams(recipient2.getAddress(), 20, 0),
                new CurrencyTypes.TransferParams(recipient3.getAddress(), 30, 0)
            );

            List<CurrencyTypes.CurrencyTransaction> txns = CurrencyTransaction.createCurrencyTransactionBatch(
                transfers,
                keyPair.getPrivateKey(),
                lastRef
            );

            assertEquals(3, txns.size());
            assertEquals(1000000000L, txns.get(0).getValue().getAmount()); // 10 * 1e8
            assertEquals(2000000000L, txns.get(1).getValue().getAmount()); // 20 * 1e8
            assertEquals(3000000000L, txns.get(2).getValue().getAmount()); // 30 * 1e8

            // Check parent references are chained
            assertEquals(lastRef, txns.get(0).getValue().getParent());
            assertEquals(6, txns.get(1).getValue().getParent().getOrdinal());
            assertEquals(7, txns.get(2).getValue().getParent().getOrdinal());
        }
    }

    @Nested
    @DisplayName("Transaction verification")
    class TransactionVerification {

        @Test
        @DisplayName("verifies correct signatures")
        void verifiesCorrectSignatures() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            CurrencyTypes.CurrencyTransaction tx = CurrencyTransaction.createCurrencyTransaction(
                new CurrencyTypes.TransferParams(keyPair2.getAddress(), 100, 0),
                keyPair.getPrivateKey(),
                lastRef
            );

            Types.VerificationResult result = CurrencyTransaction.verifyCurrencyTransaction(tx);

            assertTrue(result.isValid());
            assertEquals(1, result.getValidProofs().size());
            assertEquals(0, result.getInvalidProofs().size());
        }

        @Test
        @DisplayName("detects invalid signatures")
        void detectsInvalidSignatures() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            CurrencyTypes.CurrencyTransaction tx = CurrencyTransaction.createCurrencyTransaction(
                new CurrencyTypes.TransferParams(keyPair2.getAddress(), 100, 0),
                keyPair.getPrivateKey(),
                lastRef
            );

            // Replace with invalid proof
            Types.SignatureProof invalidProof = new Types.SignatureProof(
                tx.getProofs().get(0).getId(),
                "invalid_signature"
            );

            CurrencyTypes.CurrencyTransaction invalidTx = new CurrencyTypes.CurrencyTransaction(
                tx.getValue(),
                Arrays.asList(invalidProof)
            );

            Types.VerificationResult result = CurrencyTransaction.verifyCurrencyTransaction(invalidTx);

            assertFalse(result.isValid());
            assertEquals(0, result.getValidProofs().size());
            assertEquals(1, result.getInvalidProofs().size());
        }
    }

    @Nested
    @DisplayName("Multi-signature support")
    class MultiSignatureSupport {

        @Test
        @DisplayName("adds additional signature")
        void addsAdditionalSignature() {
            Types.KeyPair keyPair1 = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            Types.KeyPair recipient = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            // Create transaction with first signature
            CurrencyTypes.CurrencyTransaction tx = CurrencyTransaction.createCurrencyTransaction(
                new CurrencyTypes.TransferParams(recipient.getAddress(), 100, 0),
                keyPair1.getPrivateKey(),
                lastRef
            );

            assertEquals(1, tx.getProofs().size());

            // Add second signature
            tx = CurrencyTransaction.signCurrencyTransaction(tx, keyPair2.getPrivateKey());

            assertEquals(2, tx.getProofs().size());

            // Verify both signatures
            Types.VerificationResult result = CurrencyTransaction.verifyCurrencyTransaction(tx);

            assertTrue(result.isValid());
            assertEquals(2, result.getValidProofs().size());
            assertEquals(0, result.getInvalidProofs().size());
        }
    }

    @Nested
    @DisplayName("Transaction hashing")
    class TransactionHashing {

        @Test
        @DisplayName("produces consistent hashes")
        void producesConsistentHashes() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            CurrencyTypes.CurrencyTransaction tx = CurrencyTransaction.createCurrencyTransaction(
                new CurrencyTypes.TransferParams(keyPair2.getAddress(), 100, 0),
                keyPair.getPrivateKey(),
                lastRef
            );

            Types.Hash hash1 = CurrencyTransaction.hashCurrencyTransaction(tx);
            Types.Hash hash2 = CurrencyTransaction.hashCurrencyTransaction(tx);

            assertEquals(hash1.getValue(), hash2.getValue());
            assertEquals(64, hash1.getValue().length()); // SHA-256 hex string
            assertEquals(32, hash1.getBytes().length); // 32 bytes
        }

        @Test
        @DisplayName("creates correct transaction reference")
        void createsCorrectTransactionReference() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            CurrencyTypes.CurrencyTransaction tx = CurrencyTransaction.createCurrencyTransaction(
                new CurrencyTypes.TransferParams(keyPair2.getAddress(), 100, 0),
                keyPair.getPrivateKey(),
                lastRef
            );

            CurrencyTypes.TransactionReference ref = CurrencyTransaction.getTransactionReference(tx, 1);

            assertEquals(1, ref.getOrdinal());
            assertEquals(64, ref.getHash().length());
        }

        @Test
        @DisplayName("encodes transaction to string")
        void encodesTransactionToString() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            CurrencyTypes.TransactionReference lastRef = new CurrencyTypes.TransactionReference(
                "a".repeat(64), 0
            );

            CurrencyTypes.CurrencyTransaction tx = CurrencyTransaction.createCurrencyTransaction(
                new CurrencyTypes.TransferParams(keyPair2.getAddress(), 100, 0),
                keyPair.getPrivateKey(),
                lastRef
            );

            String encoded = CurrencyTransaction.encodeCurrencyTransaction(tx);

            assertNotNull(encoded);
            assertTrue(encoded.length() > 0);
        }
    }
}
